### ISSUE TITLE
Can anyone add documentation on coursier with scala native?

### ISSUE THREAD

**Issue by wisechengyi** [coursier/coursier#828]
@alexarchambault @densh maybe?

For example how to make a native binary with the current codebase and such.

**Comment by wisechengyi** [comment]
Thanks.

It fails as follows. Should I brew install anything?
```
$ cli/target/pack/bin/coursier bootstrap -S -o native-echo io.get-coursier:echo_native0.3_2.11:1.0.1
https://repo1.maven.org/maven2/io/get-coursier/echo_native0.3_2.11/1.0.1/echo_native0.3_2.11-1.0.1.pom
  100.0% [##########] 2.5 KiB (5.5 KiB / s)
https://repo1.maven.org/maven2/org/scala-native/javalib_native0.3_2.11/0.3.6/javalib_native0.3_2.11-0.3.6.pom
  100.0% [##########] 1.9 KiB (123.7 KiB / s)
https://repo1.maven.org/maven2/org/scala-native/nativelib_native0.3_2.11/0.3.6/nativelib_native0.3_2.11-0.3.6.pom
  100.0% [##########] 1.9 KiB (77.1 KiB / s)
https://repo1.maven.org/maven2/org/scala-lang/scala-library/2.11.12/scala-library-2.11.12.pom
  100.0% [##########] 1.4 KiB (69.7 KiB / s)
https://repo1.maven.org/maven2/org/scala-native/auxlib_native0.3_2.11/0.3.6/auxlib_native0.3_2.11-0.3.6.pom
  100.0% [##########] 1.9 KiB (44.1 KiB / s)
https://repo1.maven.org/maven2/org/scala-native/scalalib_native0.3_2.11/0.3.6/scalalib_native0.3_2.11-0.3.6.pom
  100.0% [##########] 2.2 KiB (38.4 KiB / s)
https://repo1.maven.org/maven2/org/scala-lang/scala-reflect/2.11.12/scala-reflect-2.11.12.pom
  100.0% [##########] 1.5 KiB (170.7 KiB / s)
https://repo1.maven.org/maven2/io/get-coursier/echo_native0.3_2.11/1.0.1/echo_native0.3_2.11-1.0.1.jar
  100.0% [##########] 6.7 KiB (69.0 KiB / s)
https://repo1.maven.org/maven2/org/scala-native/auxlib_native0.3_2.11/0.3.6/auxlib_native0.3_2.11-0.3.6.jar
  100.0% [##########] 66.6 KiB (680.1 KiB / s)
https://repo1.maven.org/maven2/org/scala-lang/scala-reflect/2.11.12/scala-reflect-2.11.12.jar
  100.0% [##########] 4.4 MiB (13.8 MiB / s)
https://repo1.maven.org/maven2/org/scala-lang/scala-library/2.11.12/scala-library-2.11.12.jar
  100.0% [##########] 5.5 MiB (19.0 MiB / s)
https://repo1.maven.org/maven2/org/scala-native/javalib_native0.3_2.11/0.3.6/javalib_native0.3_2.11-0.3.6.jar
  100.0% [##########] 2.5 MiB (5.4 MiB / s)
https://repo1.maven.org/maven2/org/scala-native/scalalib_native0.3_2.11/0.3.6/scalalib_native0.3_2.11-0.3.6.jar
  100.0% [##########] 12.2 MiB (25.3 MiB / s)
https://repo1.maven.org/maven2/org/scala-native/nativelib_native0.3_2.11/0.3.6/nativelib_native0.3_2.11-0.3.6.jar
  100.0% [##########] 849.4 KiB (1.5 MiB / s)
List(/usr/bin/clang, -S, -xc, -emit-llvm, -o, /Users/yic/workspace/coursier_3/native-target/target/ll.probe, /Users/yic/workspace/coursier_3/native-target/target/c.probe)
7 files in classpath:
Linking
Discovered 1291 classes and 9538 methods
Optimizing
Generating intermediate code
Produced 40 files
Compiling to native code
/Users/yic/workspace/coursier_3/native-target/lib/gc/boehm/gc.c:1:10: fatal error: 'gc.h' file not found
#include <gc.h>
         ^~~~~~
1 error generated.
Exception in thread "main" java.lang.RuntimeException: Failed to compile native library runtime code.
	at scala.sys.package$.error(package.scala:27)
	at coursier.extra.Native$.$anonfun$create$19(Native.scala:422)
	at scala.collection.parallel.mutable.ParArray$Map.leaf(ParArray.scala:656)
	at scala.collection.parallel.Task.$anonfun$tryLeaf$1(Tasks.scala:49)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at scala.util.control.Breaks$$anon$1.catchBreak(Breaks.scala:63)
	at scala.collection.parallel.Task.tryLeaf(Tasks.scala:52)
	at scala.collection.parallel.Task.tryLeaf$(Tasks.scala:46)
	at scala.collection.parallel.mutable.ParArray$Map.tryLeaf(ParArray.scala:647)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.compute(Tasks.scala:149)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.compute$(Tasks.scala:145)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:440)
	at java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:189)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool.helpStealer(ForkJoinPool.java:1958)
	at java.util.concurrent.ForkJoinPool.awaitJoin(ForkJoinPool.java:2047)
	at java.util.concurrent.ForkJoinTask.doJoin(ForkJoinTask.java:390)
	at java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:719)
	at scala.collection.parallel.ForkJoinTasks$WrappedTask.sync(Tasks.scala:375)
	at scala.collection.parallel.ForkJoinTasks$WrappedTask.sync$(Tasks.scala:375)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.sync(Tasks.scala:440)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.internal(Tasks.scala:170)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.internal$(Tasks.scala:153)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.internal(Tasks.scala:440)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.compute(Tasks.scala:146)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.compute$(Tasks.scala:145)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:440)
	at java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:189)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

**Comment by wisechengyi** [comment]
Follow https://github.com/crystal-lang/crystal/wiki/All-required-libraries#mac-osx-with-homebrew, so the echo one works. Cli has:

```
$ cli/target/pack/bin/coursier bootstrap -S -o native_cli io.get-coursier:coursier-cli_2.12:1.1.0-M1
https://repo1.maven.org/maven2/io/get-coursier/coursier-cli_2.12/1.1.0-M1/coursier-cli_2.12-1.1.0-M1.jar
  100.0% [##########] 460.5 KiB (1.1 MiB / s)
List(/usr/bin/clang, -S, -xc, -emit-llvm, -o, /Users/yic/workspace/coursier_3/native-target/target/ll.probe, /Users/yic/workspace/coursier_3/native-target/target/c.probe)
18 files in classpath:
Linking
cannot link: @coursier.cli.Coursier$::main_scala.scalanative.runtime.ObjectArray_unit
cannot link: @java.lang.NoSuchMethodException
cannot link: @java.lang.NoSuchMethodException::init_java.lang.String
cannot link: @java.lang.Object::equals_java.lang.Object_bool
cannot link: @java.lang.Object::hashCode_i32
cannot link: @java.lang.Object::scala$underscore$##_i32
cannot link: @java.lang.Object::scala$underscore$==_java.lang.Object_bool
cannot link: @java.lang.String
cannot link: @java.lang.String::field.cachedHashCode
cannot link: @java.lang.String::field.count
cannot link: @java.lang.String::field.offset
cannot link: @java.lang.String::field.value
cannot link: @java.lang.Throwable::printStackTrace_unit
cannot link: @scala.runtime.BoxesRunTime$
cannot link: @scala.runtime.BoxesRunTime$::boxToBoolean_bool_java.lang.Boolean
cannot link: @scala.runtime.BoxesRunTime$::boxToByte_i8_java.lang.Byte
cannot link: @scala.runtime.BoxesRunTime$::boxToCharacter_char_java.lang.Character
cannot link: @scala.runtime.BoxesRunTime$::boxToDouble_f64_java.lang.Double
cannot link: @scala.runtime.BoxesRunTime$::boxToFloat_f32_java.lang.Float
cannot link: @scala.runtime.BoxesRunTime$::boxToInteger_i32_java.lang.Integer
cannot link: @scala.runtime.BoxesRunTime$::boxToLong_i64_java.lang.Long
cannot link: @scala.runtime.BoxesRunTime$::boxToShort_i16_java.lang.Short
cannot link: @scala.runtime.BoxesRunTime$::unboxToBoolean_java.lang.Object_bool
cannot link: @scala.runtime.BoxesRunTime$::unboxToByte_java.lang.Object_i8
cannot link: @scala.runtime.BoxesRunTime$::unboxToChar_java.lang.Object_char
cannot link: @scala.runtime.BoxesRunTime$::unboxToDouble_java.lang.Object_f64
cannot link: @scala.runtime.BoxesRunTime$::unboxToFloat_java.lang.Object_f32
cannot link: @scala.runtime.BoxesRunTime$::unboxToInt_java.lang.Object_i32
cannot link: @scala.runtime.BoxesRunTime$::unboxToLong_java.lang.Object_i64
cannot link: @scala.runtime.BoxesRunTime$::unboxToShort_java.lang.Object_i16
cannot link: @scala.scalanative.runtime.BoxedUnit$
cannot link: @scala.scalanative.runtime.Boxes$
cannot link: @scala.scalanative.runtime.Boxes$::boxToUByte_i8_java.lang.Object
cannot link: @scala.scalanative.runtime.Boxes$::boxToUInt_i32_java.lang.Object
cannot link: @scala.scalanative.runtime.Boxes$::boxToULong_i64_java.lang.Object
cannot link: @scala.scalanative.runtime.Boxes$::boxToUShort_i16_java.lang.Object
cannot link: @scala.scalanative.runtime.Boxes$::unboxToUByte_java.lang.Object_i8
cannot link: @scala.scalanative.runtime.Boxes$::unboxToUInt_java.lang.Object_i32
cannot link: @scala.scalanative.runtime.Boxes$::unboxToULong_java.lang.Object_i64
cannot link: @scala.scalanative.runtime.Boxes$::unboxToUShort_java.lang.Object_i16
cannot link: @scala.scalanative.runtime.CharArray
cannot link: @scala.scalanative.runtime.ObjectArray
cannot link: @scala.scalanative.runtime.package$
cannot link: @scala.scalanative.runtime.package$::init_i32_ptr_scala.scalanative.runtime.ObjectArray
cannot link: @scala.scalanative.runtime.package$::loop_unit
Exception in thread "main" java.lang.RuntimeException: unable to link
	at scala.sys.package$.error(package.scala:27)
	at coursier.extra.Native$.create(Native.scala:309)
	at coursier.cli.Bootstrap$.run(Bootstrap.scala:64)
	at coursier.cli.Coursier$.$anonfun$runA$2(Coursier.scala:16)
	at coursier.cli.Coursier$.$anonfun$runA$2$adapted(Coursier.scala:14)
	at caseapp.core.app.CommandAppA.run(CommandAppA.scala:16)
	at caseapp.core.app.CommandAppWithPreCommand.$anonfun$main$5(CommandAppWithPreCommand.scala:98)
	at caseapp.core.app.CommandAppWithPreCommand.$anonfun$main$5$adapted(CommandAppWithPreCommand.scala:98)
	at scala.util.Either.fold(Either.scala:188)
	at caseapp.core.app.CommandAppWithPreCommand.$anonfun$main$3(CommandAppWithPreCommand.scala:98)
	at caseapp.core.app.CommandAppWithPreCommand.$anonfun$main$3$adapted(CommandAppWithPreCommand.scala:84)
	at scala.Option.foreach(Option.scala:257)
	at caseapp.core.app.CommandAppWithPreCommand.main(CommandAppWithPreCommand.scala:84)
	at coursier.cli.Coursier.main(Coursier.scala)
```

**Comment by alexarchambault** [comment]
Yeah, coursier itself isn't compiled for scala native… It can only generate native binaries, using the scala-native tool chain that runs on the JVM (like the scala-native sbt plugin does).

The issue to have coursier itself compiled with scala-native is [there](https://github.com/coursier/coursier/issues/458). Last time I checked, scala-native wasn't supposed to support multi-threading, but handling it via a purely native library like libuv might work. I tried to do some basic stuff with libuv from scala-native, but ran into a few issues… (I'd like to clean them up a bit and push that to get advices…)
